Extend the mesh connectivity section (#231)
precice/precice.github.io: pages/docs/couple-your-code/couple-your-code-defining-mesh-connectivity.md

The most important example where mesh connectivity is needed is the `nearest-projection` mapping, where the mesh we project _into_ needs mesh connectivity. For a consistent mapping, this is the mesh _from_ which you map. For a conservative mapping, the mesh _to_ which you map. More information is given on the [mapping configuration page](configuration-mapping).

There are two types of connectivity, which depend on the type of coupling.
For surface coupling in 2D, mesh connectivity boils down to defining edges between vertices. In 3D, you need to define triangles and / or quads.
For volume coupling in 2D, mesh connectivity boils down to defining triangles and / or quads between vertices. In 3D, you need to define tetrahedra introduced in version `2.5.0`.

> All kind of connectivity can be either build up directly from vertices. Triangles and Quads also allow you to define them using edge IDs.

Reviewer: ```suggestion
All kind of connectivity can be built up directly from vertices. Triangles and quads also allow us to define them using edge IDs.
```

void setMeshTriangleWithEdges (int meshID, int firstVertexID, int secondVertexID, int thirdVertexID);
void setMeshQuad(int meshID, int firstEdgeID, int secondEdgeID, int thirdEdgeID, int fourthEdgeID);
void setMeshQuadWithEdges(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
void setMeshTetrahredron(int meshID, int firstVertexID, int secondVertexID, int thirdVertexID, int fourthVertexID);
```

* `setMeshEdge` defines a mesh edge between two vertices and returns an edge ID.
* `setMeshTriangle` defines a mesh triangle by three edges.
* `setMeshTriangleWithEdges` defines a mesh triangle by three vertices and also creates the edges in preCICE on the fly. Of course, preCICE takes care that no edge is defined twice. Please note that this function is computationally more expensive than `setMeshTriangle`.
* `setMeshQuad` defines a mesh quad by four edges.
* `setMeshQuadWithEdges` defines a mesh quad by four vertices and also creates the edges in preCICE on the fly. Again, preCICE takes care that no edge is defined twice. This function is computationally more expensive than `setMeshQuad`.

> * `setMeshTetrahredron` defines a mesh tetrahedra by four vertices.

Reviewer: ```suggestion
* `setMeshTetrahredron` defines a mesh tetrahedron by four vertices.
```

[...]

```

## Changes in v3

Version 3 overhauls the definition of meshes.

> Connectivity now consists of explicitly defined elements (via the API) and implicitly defined elements.
> preCICE automatically generates all implicit elements.

Reviewer: ```suggestion
Connectivity now consists of explicitly defined elements (elements created via calls to the API) and implicitly defined elements (elements additionally created by preCICE).
```

As an example, explicitly defining a triangle ABC via the API guarantees the existence of the implicit edges AB, AC, and BC.

> Furthermore, all connectivity is defined via vertex IDs only. There are no more edge IDs to worry about.

Reviewer: ```suggestion
Furthermore, all connectivity is defined only via vertex IDs. There are no more edge IDs to worry about.
```

> The order of vertices also doesn't matter. Triangles BAC and CAB are considered duplicates.

Reviewer: I assume this is what "considered duplicates" means:

```suggestion
The order of vertices also does not matter. Triangles BAC and CAB are considered duplicates and preCICE will ignore additional definitions of the same triangle.
```

(somehow the `does not` is supposed to be clearer than `doesn't`)

Author: The addition is not true. preCICE has to remove duplicates, which is computationally expensive. We should clarify that this comes at a cost.

```suggestion
The order of vertices also does not matter. Triangles BAC and CAB are considered duplicates and preCICE removes one of them during the deduplication step.
```

> ```cpp

Reviewer: A sentence introducing this list (or an explicit connection to the previous one) is missing.

Author: ```suggestion
The API for defining individual connectivity elements looks as follows:

```cpp
```
